spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -592,120 +606,575 @@ def _write_synaptic_matrix_and_master_population_table(
+        return block_addr, syn_mat_addr
+
+    def __write_machine_matrix(
+            self, block_addr, single_addr, spec,
+            max_synapses, max_words, r_info, row_data, synapse_info, pre_slice,
+            post_vertex_slice, single_synapses, all_syn_block_sz, is_delayed):
+        # If there are no synapses, don't write anything
+        if max_synapses == 0:
+            # If there is routing information, write an invalid entry
+            if r_info is not None:
+                self.__poptable_type.add_invalid_entry(
+                    r_info.first_key_and_mask, 0, 0, 0)
+            return block_addr, single_addr, None, None
+
+        # Write a matrix for an incoming machine vertex
+        if max_synapses == 1 and self.__is_direct(
+                single_addr, synapse_info, pre_slice, post_vertex_slice,
+                is_delayed):
+            single_rows = row_data.reshape(-1, 4)[:, 3]
+            index = self.__poptable_type.update_master_population_table(
+                single_addr, max_words, r_info.first_key_and_mask, 0, 0, 0,
+                is_single=True)
+            self.__update_synapse_index(synapse_info, post_vertex_slice, index)
+            single_synapses.append(single_rows)
+            syn_mat_offset = single_addr
+            single_addr = single_addr + (len(single_rows) * 4)
+            return block_addr, single_addr, syn_mat_offset, True
+        block_addr = self._write_pop_table_padding(spec, block_addr)
+        index = self.__poptable_type.update_master_population_table(
+            block_addr, max_words, r_info.first_key_and_mask, 0, 0, 0)
+        self.__update_synapse_index(synapse_info, post_vertex_slice, index)
+        spec.write_array(row_data)
+        syn_mat_offset = block_addr
+        block_addr = block_addr + (len(row_data) * 4)
+        if block_addr > all_syn_block_sz:
+            raise Exception(
+                "Too much synaptic memory has been written: {} of {} "
+                .format(block_addr, all_syn_block_sz))
+        return block_addr, single_addr, syn_mat_offset, False
+
+    @staticmethod
+    def __check_keys_adjacent(keys, mask_size):
+        # Check that keys are all adjacent
+        key_increment = (1 << mask_size)
+        last_key = None
+        last_slice = None
+        for i, (key, v_slice) in enumerate(keys):
+            if last_key is None:
+                last_key = key
+                last_slice = v_slice
+            elif (last_key + key_increment) != key:
+                return False
+            elif (i + 1) < len(keys) and last_slice.n_atoms != v_slice.n_atoms:
+                return False
+            elif (last_slice.hi_atom + 1) != v_slice.lo_atom:
+                return False
+            last_key = key
+            last_slice = v_slice
+        return True
+
+    def __get_app_key_and_mask(self, keys, mask, n_stages, key_space_tracker):
+
+        # Can be merged only if keys are adjacent outside the mask
```

Comment:
misses the chekc that all slices are the same slice.

Reply:
Same *size* now added